### PR TITLE
[docs] revamp documentation

### PR DESCRIPTION
**Pull request by tscholak**
# ✨ Description

The documentation is in dire need of improvements... this is just the beginning.

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [ ] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [x] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

- [x] add docs `README.md`: https://github.com/ServiceNow/Fast-LLM/tree/tscholak/improve-docs/docs
- [x] add new landing page: https://github.com/ServiceNow/Fast-LLM/blob/tscholak/improve-docs/docs/index.md
- [x] add single-node quick start guide: https://github.com/ServiceNow/Fast-LLM/blob/tscholak/improve-docs/docs/quick-start.md
- [x] add new distributed data preparation script: https://github.com/ServiceNow/Fast-LLM/blob/tscholak/improve-docs/tools/prepare_dataset.py
- [x] add cost efficiency comparison: https://github.com/ServiceNow/Fast-LLM/blob/tscholak/improve-docs/docs/cost-efficiency.md
    - [ ] measure throughput for the different frameworks
- [x] add general help page: https://github.com/ServiceNow/Fast-LLM/blob/tscholak/improve-docs/docs/help.md
- [x] add StarCoder 2 success story: https://github.com/ServiceNow/Fast-LLM/blob/tscholak/improve-docs/docs/success-stories/starcoder-2.md
- [x] add about us section: https://github.com/ServiceNow/Fast-LLM/blob/tscholak/improve-docs/docs/about-us.md
- [x] add join us page: https://github.com/ServiceNow/Fast-LLM/blob/tscholak/improve-docs/docs/join-us.md

- [ ] add Fast-LLM in action section:
    - [ ] add Slurm guide (dockerized and non-dockerized)
    - [ ] add Kubernetes guide (with Kubeflow)
- [ ] add real-world examples:
    - [ ] add training data preparation
        - [ ] add DCLM preparation
        - [ ] add The Stack preparation
        - [ ] add weighted mix preparation
    - [ ] add train Llama 8B from scratch guide
    - [ ] add Llama 8B continual pretraining guide
    - [ ] add upcycle Llama 3B to MoE guide
- [ ] add reference section:
    - [ ] add auto-generated configuration option reference.
- [ ] add developer's corner:
    - [ ] add contributing guidelines
    - [ ] add best development practices
        - [ ] describe setting up a dev environment
        - [ ] describe launching test jobs
        - [ ] describe debugging
        - [ ] describe profiling

# ✅ Checklist

Make sure the following tasks are completed before submitting the PR:

### General:
- [ ] 📜 I have read and followed the [contributing guidelines](CONTRIBUTING.md).
- [ ] 🎉 The functionality is complete, and I have tested the changes.
- [x] 📝 I have updated the documentation if needed.
- [ ] ⚠️ The change does not introduce any new issues (e.g., runtime warnings, type checker errors, linting problems, unhandled edge cases).
- [ ] 🧩 I have commented my code, especially in hard-to-understand areas.

### Dependencies and Configuration:
- [ ] 🐋 I have updated the Docker configuration or dependencies, if applicable.
- [ ] 🔄 I have ensured compatibility with the existing setup after dependency changes.

### Testing:
- [ ] 🧪 I have added or updated tests to cover my changes.
- [ ] ✔️ New and existing tests pass locally with my changes.
- [ ] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact:
- [ ] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

# 📊 Performance Impact Details

If there is any impact on performance, describe it and provide benchmark results, if applicable:

---

# 📝 Additional Notes

Include any additional context, information, or considerations here, such as known issues, follow-up tasks, or backward compatibility concerns.

### CODE REVIEW

File: docs/join-us.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+---
+title: Join Us
+hide:
+  - navigation
+---
+
+Fast-LLM is an open-source project driven by a community of passionate contributors. Whether you're a researcher, developer, or AI enthusiast, there's a place for you to make a real impact on the future of large-scale AI training. Join us, dive in, and help shape the tools that push the boundaries of language model training. Here's how you can get involved:
+
+---
+
```
Comment (jlamypoirier) [inline docs/join-us.md:10]: Should we add a servicenow recruitement ad?

File: docs/blog/index.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Blog
```
Comment (jlamypoirier) [inline docs/blog/index.md:1]: Is this used at all?

File: docs/refs.bib [hunk 3]
```diff
@@ -0,0 +1,13 @@
+@article{li2023starcoder,
```
Comment (jlamypoirier) [inline docs/refs.bib:1]: Where does this show?

File: docs/refs.bib [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  journal={arXiv preprint arXiv:2305.06161},
+  year={2023}
+}
+
+@article{lozhkov2024starcoder,
+  title={Starcoder 2 and the stack v2: The next generation},
+  author={Lozhkov, Anton and Li, Raymond and Allal, Loubna Ben and Cassano, Federico and Lamy-Poirier, Joel and Tazi, Nouamane and Tang, Ao and Pykhtar, Dmytro and Liu, Jiawei and Wei, Yuxiang and others},
+  journal={arXiv preprint arXiv:2402.19173},
+  year={2024}
+}
```
Comment (jlamypoirier) [inline docs/refs.bib:13]: Missing newline. Did you setup pre-commit?

File: docs/success-stories.md
Comment (jlamypoirier) [inline docs/success-stories.md:1]: Is this file needed?

**Review by jlamypoirier — COMMENTED** [review]
First part of the review, excludes `examples`, `get started` and `developers` sections.
Overall looks great except for some minor details

File: tools/prepare_dataset.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    @cached_property
+    def tokenizer(self):
+        return AutoTokenizer.from_pretrained(self.tokenizer_path_or_name)
+
+    def _validate(self):
+        if self.dataset_dtype is None:
+            # Decide the dtype based on the tokenizer vocabulary size
+            vocab_size = len(self.tokenizer)
+
+            if vocab_size <= np.iinfo(np.int8).max:
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:105]: I don't think int8 (would need to be uint8) or int64 are realistic

File: tools/prepare_dataset.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        desc="Total number of processes in distributed processing.",
+        hint=FieldHint.optional,
+    )
+    distributed_backend: str = Field(
+        default="gloo",
+        desc="Distributed backend for distributed processing.",
+        hint=FieldHint.optional,
+    )
+
+    @cached_property
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:96]: Probably not compatible with `Config`. Usually I use a Field(init=False) and set in validation.

File: tools/prepare_dataset.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+logging.set_verbosity_error()
+
+
+@config_class
+class PrepareDatasetConfig(RunnableConfig):
+    dataset_name_or_path: str = Field(
+        desc="Name or path of the dataset.",
+        hint=FieldHint.core,
+    )
+    tokenizer_path_or_name: str = Field(
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:28]: How about Fast-LLM's TokenizerConfig and Tokenizer?

File: tools/prepare_dataset.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+from tqdm import tqdm
+from transformers import AutoTokenizer, logging
+
+from fast_llm.config import Field, FieldHint, check_field, config_class
+from fast_llm.data.mmap import MMapIndexedDataset
+from fast_llm.engine.config_utils.data_type import DataType
+from fast_llm.engine.config_utils.runnable import RunnableConfig
+from fast_llm.utils import Assert
+
+logging.set_verbosity_error()
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:19]: Why?
Should be in `if __name__ == "__main__":` at least

File: tools/prepare_dataset.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
+import json
+import os
+from functools import cached_property
+from multiprocessing import Pool
+from pathlib import Path
+
+import numpy as np
+from datasets import load_dataset, load_from_disk
+from torch import distributed as dist
+from tqdm import tqdm
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:10]: Not in the requirements. Not sure we need it, but if really needed you'll need to add it to optional requirements.

File: tools/prepare_dataset.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
+import json
+import os
+from functools import cached_property
+from multiprocessing import Pool
+from pathlib import Path
+
+import numpy as np
+from datasets import load_dataset, load_from_disk
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:8]: needs to be added to the requirements

File: tools/prepare_dataset.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+from torch import distributed as dist
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:9]: Fast-LLM coding convention. Import all first-party names with `as`, all other without.

File: tools/prepare_dataset.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+            "prefix": prefix,
+            "num_documents": len(documents),
+            "num_tokens": sum(len(doc) for doc in documents),
+        }
+        return dataset_dict
+
+    def run(self):
+        # Initialize distributed processing
+        if self.world_size > 1:
+            dist.init_process_group(backend=self.distributed_backend, rank=self.rank, world_size=self.world_size)
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:155]: Let's use `Distributed` here, and `DistributedConfig` in the config

File: tools/prepare_dataset.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+    )
+    tokenizer_path_or_name: str = Field(
+        desc="Path or name of the tokenizer.",
+        hint=FieldHint.core,
+    )
+    output_dir: str = Field(
+        desc="Output directory for the processed dataset.",
+        hint=FieldHint.core,
+    )
+    num_processes_load: int = Field(
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:36]: Naming convention for fields:
* No abbreviations
* As self-explanatory as possible
* Avoid redundancies (ex. num)

File: tools/prepare_dataset.py [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+class PrepareDatasetConfig(RunnableConfig):
+    dataset_name_or_path: str = Field(
+        desc="Name or path of the dataset.",
+        hint=FieldHint.core,
+    )
+    tokenizer_path_or_name: str = Field(
+        desc="Path or name of the tokenizer.",
+        hint=FieldHint.core,
+    )
+    output_dir: str = Field(
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:32]: output_path: pathlib.Path

File: tools/prepare_dataset.py (hunk 13 above)
Comment (jlamypoirier) [reply tools/prepare_dataset.py:36]: Here `loading_workers` I think?

File: tools/prepare_dataset.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+        desc="Output directory for the processed dataset.",
+        hint=FieldHint.core,
+    )
+    num_processes_load: int = Field(
+        default=1,
+        desc="Number of workers in load_dataset() call.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 0),
+    )
+    num_processes_map: int = Field(
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:42]: `mapping_workers`?

File: tools/prepare_dataset.py [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+        default=1,
+        desc="Number of workers in .map() call.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 0),
+    )
+    num_processes_save: int = Field(
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:48]: `saving_workers`?

File: tools/prepare_dataset.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 0),
+    )
+    num_tokens_per_shard: int = Field(
+        default=1000000000,
+        desc="Approximate number of tokens per shard.",
+        hint=FieldHint.optional,
+        valid=check_field(Assert.geq, 1),
+    )
+    dataset_config_name: None | str = Field(
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:60]: What does this do?

File: tools/prepare_dataset.py [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+        default=None,
+        desc="Specific configuration name for the dataset.",
+        hint=FieldHint.optional,
+    )
+    dataset_split: str = Field(
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:65]: Consider making a sub-config for input dataset to avoid all the `dataset` prefixes?

File: tools/prepare_dataset.py [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+        default="train",
+        desc="Split of the dataset to use.",
+        hint=FieldHint.optional,
+    )
+    dataset_field: str = Field(
+        default="text",
+        desc="Field of the dataset to use.",
+        hint=FieldHint.optional,
+    )
+    dataset_dtype: DataType = Field(
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:75]: `data_type`

File: tools/prepare_dataset.py (hunk 18 above)
Comment (jlamypoirier) [reply tools/prepare_dataset.py:65]: Also what does the split do?

File: tools/prepare_dataset.py [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+        }
+        return dataset_dict
+
+    def run(self):
+        # Initialize distributed processing
+        if self.world_size > 1:
+            dist.init_process_group(backend=self.distributed_backend, rank=self.rank, world_size=self.world_size)
+
+        # Prepare output directory
+        os.makedirs(self.output_dir, exist_ok=True)
```
Comment (jlamypoirier) [inline tools/prepare_dataset.py:158]: Fast-LLM convention: use pathlib (field should be a Path already)

File: docs/refs.bib (hunk 3 above)
Comment (tscholak) [reply docs/refs.bib:1]: StarCoder 2 success story

File: docs/quick-start.md [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+  wandb:
+    project_name: fast-llm
+    entity_name: servicenow  # (2)!
+    tags: quick-start
+    alert:
+      interval: 1000
+batch:
+  micro_batch_size: 1  # (3)!
+  sequence_length: 1024  # (4)!
+  batch_size: 480  # (5)!
```
[the remark this answers is not in the document]
Comment (tscholak) [reply docs/quick-start.md:117]: Got that number from Kaparthy's tutorial

File: tools/prepare_dataset.py (hunk 18 above)
Comment (tscholak) [reply tools/prepare_dataset.py:65]: hf datasets come with multiple splits. we need to select one, and that is usually `train`.

File: tools/prepare_dataset.py (hunk 17 above)
Comment (tscholak) [reply tools/prepare_dataset.py:60]: some datasets cannot be used unless a dataset config is specified that is then used to grab a selection of the dataset.

**Comment by tscholak** [comment]
Hi, we need to move forward with this. Let's leave improvements and corrections for these docs for future PRs.
